common/data_refinery_common/models/batches.py

```diff
@@ -0,0 +1,277 @@
+import logging
+logger = logging.getLogger(__name__)
+
+
+DEFAULT_BATCH_PREFIX = "batch_"
+RAW_PREFIX = get_env_variable("RAW_PREFIX")
+TEMP_PREFIX = get_env_variable("TEMP_PREFIX")
+PROCESSED_PREFIX = get_env_variable("PROCESSED_PREFIX")
+LOCAL_ROOT_DIR = get_env_variable("LOCAL_ROOT_DIR")
+USE_S3 = get_env_variable("USE_S3") == "True"
+S3_BUCKET_NAME = get_env_variable("S3_BUCKET_NAME")
+
+
+class BatchStatuses(Enum):
+    """Valid values for the status field of the Batch model."""
+
+    NEW = "NEW"
+    DOWNLOADED = "DOWNLOADED"
+    PROCESSED = "PROCESSED"
+
+
+class Batch(TimeTrackedModel):
+    """Represents a batch of data.
+
+    The definition of a Batch is intentionally that vague. What a batch
+    is will vary from source to source. It could be a single file, or
+    a group of files with some kind of logical grouping such as an
+    experiment.
+    """
+
+    survey_job = models.ForeignKey(SurveyJob, on_delete=models.PROTECT)
+    source_type = models.CharField(max_length=256)
+    pipeline_required = models.CharField(max_length=256)
+    platform_accession_code = models.CharField(max_length=32)
+    experiment_accession_code = models.CharField(max_length=32)
+    experiment_title = models.CharField(max_length=256)
+    status = models.CharField(max_length=20)
+    release_date = models.DateField()
+    last_uploaded_date = models.DateField()
+
+    # This corresponds to the organism taxonomy ID from NCBI.
+    organism_id = models.IntegerField()
+    # This is the organism name as it appeared in the experiment.
+    organism_name = models.CharField(max_length=256)
+
+    _files = None
+
+    def _get_files(self):
+        if self._files is None:
+            return list(self.file_set.all())
+        else:
+            return self._files
+
+    def _set_files(self, new_files):
+        self._files = new_files
+
+    files = property(_get_files, _set_files)
+
+    @classmethod
+    def is_new_batch(cls, batch) -> bool:
+        file_names = batch.file_set.all().values("name")
+        matching_batches = cls.objects.filter(name__in=file_names)
+        return matching_batches.count() == 0
+
+    @classmethod
+    @transaction.atomic
+    def create_file_and_relationships(cls, *args, **kwargs):
+        """Inits and saves a batch and its relationships to its files.
+
+        Expects keyword arguments that could be passed to the init
+        method for Batch, with the addition of the keyword
+        argument 'files', which must be specified as a list of File
+        objects which have already been saved (so they have an id).
+        """
+        files = kwargs.pop('files', None)
+        this_batch = cls(*args, **kwargs)
+        if files is None:
+            raise KeyError("The 'files' argument must be specified.")
+        else:
+            this_batch.save()
+            this_batch.files.add(*files)
+
+        return this_batch
+
+    class Meta:
+        db_table = "batches"
+
+
+class File(TimeTrackedModel):
```

Comment:
So `File` is a new class that was refactored out of the original `Batch` class, am I right?

Reply:
Yes, the reason for this is that Batches should be able to have more than one File, which they can't if the fields are part of the Batch.